### PR TITLE
fix: closing editor on parsing error

### PR DESCRIPTION
**Pull request by kostagersh**
Closes CRVD-1133

### CODE REVIEW

File: packages/corvid-local-site/src/siteConverter.js [hunk 1]
```diff
@@ -199,13 +199,19 @@ const editorDocumentToLocalDocumentFiles = siteDocument => {
...
+    if (isPathOfWixFile(file.path)) {
+      try {
+        return Object.assign({}, file, {
           content: fromWixFileContent(file.content).content
-        })
-      : file
-  );
+        });
+      } catch (_) {
+        throw new Error(`Cannot parse file [${file.path}]`);
```
Comment (kostagersh) [inline packages/corvid-local-site/src/siteConverter.js]: maybe we need to add ParseError and handle it in getSiteDocument?

File: packages/corvid-cli/src/apps/open-editor.js [hunk 2]
```diff
@@ -20,6 +20,11 @@ const openEditorApp = ({ useSsl = true } = {}) => ({
         console.log(message);
       });
 
+      client.on("fatal-error", message => {
+        logger.warn(`Fatal Error ${message}`);
```
Comment (tomenden) [inline packages/corvid-cli/src/apps/open-editor.js]: In a scenario which:
1. The user changed a .wix file
2. The .wix file **is** parse-able
Will the editor report a fatal error? What will be the message the user sees?

File: packages/corvid-local-site/src/siteConverter.js (hunk 1 above)
Comment (tomenden) [reply packages/corvid-local-site/src/siteConverter.js]: Please elaborate

Comment (kostagersh) [reply packages/corvid-local-site/src/siteConverter.js]: throw new ParseError(file.path) and then catch specific type of error so it will be not some 'fatal-error', but 'parse-error' with specific message

File: packages/corvid-cli/src/apps/open-editor.js (hunk 2 above)
Comment (kostagersh) [reply packages/corvid-cli/src/apps/open-editor.js]: if user had changed .wix file in middle of a session editor will ask him to refresh and after refresh in will close electron and write parse error in cli

**Comment by tomenden** [comment]
I meant a case where the user made the change offline. In other words,
cases where the editor opens with invalid data, there are such issues in
sentry I believe (I think the editor error says something about path not
being an array or something like that

**Comment by kostagersh** [comment]
Editor will close after load and error will appear in cli (Parse error in file pages/c1dmp.wix)

File: packages/corvid-local-server/src/socket-api/index.js [hunk 3]
```diff
@@ -48,9 +48,10 @@ const initServerApi = (
   const isEditorConnected = () =>
     Object.keys(editorServer.io.sockets.connected).length > 0;
 
-  const getSiteDocument = () => {
-    return localSite.getSiteDocument();
-  };
+  const getSiteDocument = () =>
+    localSite
+      .getSiteDocument()
+      .catch(e => notifyAdmin("fatal-error", e.message));
```
Comment (dor-itzhaki) [inline packages/corvid-local-server/src/socket-api/index.js]: this does not report the error to the editor.
if we revert this change, won't the editor catch the error and report it to the cli by itself ?

File: packages/corvid-cli/src/apps/open-editor.js (hunk 2 above)
Comment (kostagersh) [reply packages/corvid-cli/src/apps/open-editor.js]: if .wix file is parse-able then electron will not close of course

File: packages/corvid-cli/src/apps/open-editor.js [hunk 4]
```diff
@@ -20,6 +20,12 @@ const openEditorApp = ({ useSsl = true } = {}) => ({
         console.log(message);
       });
 
+      client.on("parse-error", path => {
+        const message = `Parse Error in file ${path}`;
```
Comment (dor-itzhaki) [inline packages/corvid-cli/src/apps/open-editor.js]: we need to rephrase it for users i think

File: packages/corvid-local-logger/src/ParseError.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+const ExtendableError = require("es6-error");
+
+class ParseError extends ExtendableError {
```
Comment (dor-itzhaki) [inline packages/corvid-local-logger/src/ParseError.js]: This shouldn't be part of corvid-local-logger

File: packages/corvid-local-site/src/siteConverter.js [hunk 6]
```diff
@@ -199,13 +201,19 @@ const editorDocumentToLocalDocumentFiles = siteDocument => {
...
+    if (isPathOfWixFile(file.path)) {
+      try {
+        return Object.assign({}, file, {
           content: fromWixFileContent(file.content).content
-        })
-      : file
-  );
+        });
+      } catch (_) {
+        throw new ParseError(file.path);
```
Comment (dor-itzhaki) [inline packages/corvid-local-site/src/siteConverter.js]: fromWixFileContent can throw the error. wdyt ?

**Comment by dor-itzhaki** [comment]
@kostagersh we need to also make sure that the santa editor shows an understandable error when it loads with bad site document.

File: packages/corvid-local-site/src/siteConverter.js (hunk 6 above)
Comment (kostagersh) [reply packages/corvid-local-site/src/siteConverter.js]: i need file path there

File: packages/corvid-local-logger/src/ParseError.js (hunk 5 above)
Comment (kostagersh) [reply packages/corvid-local-logger/src/ParseError.js]: i can put it in utils

File: packages/corvid-local-site/src/siteConverter.js [hunk 7]
```diff
@@ -199,13 +201,19 @@ const editorDocumentToLocalDocumentFiles = siteDocument => {
...
+  const parsedLocalDocumentFiles = localDocumentFiles.map(file => {
+    if (isPathOfWixFile(file.path)) {
+      try {
+        return Object.assign({}, file, {
           content: fromWixFileContent(file.content).content
-        })
-      : file
-  );
+        });
+      } catch (_) {
```
Comment (dor-itzhaki) [inline packages/corvid-local-site/src/siteConverter.js]: we shouldn't ignore the actual original error

Comment (kostagersh) [reply packages/corvid-local-site/src/siteConverter.js]: done

File: packages/corvid-cli/src/apps/open-editor.js (hunk 4 above)
Comment (kostagersh) [reply packages/corvid-cli/src/apps/open-editor.js]: @wixysam can you help with rephrasing?

File: packages/corvid-cli/src/apps/open-editor.js [hunk 8]
```diff
@@ -61,6 +67,10 @@ const openEditorApp = ({ useSsl = true } = {}) => ({
                 getMessage("OpenEditor_Client_Console_Fatal_Error")
               )
             );
+          },
+          [clientMessages.DECODE_ERROR_MESSAGE]: path => {
+            const message = `decode error in ${path}`;
```
Comment (kostagersh) [inline packages/corvid-cli/src/apps/open-editor.js]: @wixysam and here

File: packages/corvid-cli/src/apps/open-editor.js (hunk 4 above)
Comment (wixysam) [reply packages/corvid-cli/src/apps/open-editor.js]: How do the files get to this situation? Is there something to do about it?

Comment (wixysam) [reply packages/corvid-cli/src/apps/open-editor.js]: How do the files get to this situation? Is there something to do about it?

File: packages/corvid-cli/src/apps/open-editor.js (hunk 8 above)
Comment (wixysam) [reply packages/corvid-cli/src/apps/open-editor.js]: Same questions as above.

File: packages/corvid-cli/src/apps/open-editor.js (hunk 4 above)
Comment (kostagersh) [reply packages/corvid-cli/src/apps/open-editor.js]: User actions mostly

File: packages/corvid-cli/src/apps/open-editor.js (hunk 8 above)
Comment (kostagersh) [reply packages/corvid-cli/src/apps/open-editor.js]: user actions or maybe old version of site and we had changed encoding function (shouldn't happen)

File: packages/corvid-cli/src/apps/open-editor.js (hunk 4 above)
Comment (wixysam) [reply packages/corvid-cli/src/apps/open-editor.js]: And how do they fix it? Try to revert to an old version, do a pull, anything else?

**Comment by kostagersh** [comment]
I think "Try to revert to an old version" is ok, we don't know if git is
used

File: packages/corvid-cli/src/apps/open-editor.js (hunk 4 above)
Comment (wixysam) [reply packages/corvid-cli/src/apps/open-editor.js]: I want to tag @chaimk for his opinion here, but he is not a member of this organization. Maybe turn around and give him a tap.

Here's what I have so far:

Error parsing ${path}. The file has been corrupted. Try reverting to an older version.

File: packages/corvid-cli/src/apps/open-editor.js (hunk 8 above)
Comment (wixysam) [reply packages/corvid-cli/src/apps/open-editor.js]: Same idea as above: Error decoding ${path}. The file has been corrupted. Try reverting to an older version.

Comment (kostagersh) [reply packages/corvid-cli/src/apps/open-editor.js]: there is a problem here, path is not a file and we don't even know to which file it will be converted :(. it is something like "pages.d1dmp.content"

File: packages/corvid-cli/src/apps/open-editor.js (hunk 4 above)
Comment (kostagersh) [reply packages/corvid-cli/src/apps/open-editor.js]: ok

File: packages/corvid-cli/src/apps/open-editor.js (hunk 8 above)
Comment (wixysam) [reply packages/corvid-cli/src/apps/open-editor.js]: Then leave out the middle sentence. I guess that's the best we can do.

Comment (kostagersh) [reply packages/corvid-cli/src/apps/open-editor.js]: Thanks!

File: packages/corvid-local-logger/src/ParseError.js (hunk 5 above)
Comment (kostagersh) [reply packages/corvid-local-logger/src/ParseError.js]: done

File: packages/corvid-cli/test/it/edit.spec.js [hunk 9]
```diff
@@ -255,4 +255,49 @@ describe("edit", () => {
...
+      .mock(
+        `http://frog.wix.com/code?src=39&evid=201&msid=12345678&uuid=testGuid&csi=${
+          process.env.CORVID_SESSION_ID
+        }&status=start`,
+        JSON.stringify({})
+      )
+      .mock(
+        `http://frog.wix.com/code?src=39&evid=201&msid=12345678&uuid=testGuid&csi=${
+          process.env.CORVID_SESSION_ID
+        }&status=success`,
```
Comment (dor-itzhaki) [inline packages/corvid-cli/test/it/edit.spec.js]: how can it report a success ?

File: packages/corvid-cli/test/it/edit.spec.js [hunk 10]
```diff
@@ -255,4 +255,49 @@ describe("edit", () => {
...
+        JSON.stringify({})
+      );
+
+    return expect(openEditor(tempDir)).rejects.toThrow();
```
Comment (dor-itzhaki) [inline packages/corvid-cli/test/it/edit.spec.js]: we need to check that the relevant message is shown to the user

File: packages/corvid-cli/src/apps/open-editor.js [hunk 11]
```diff
@@ -61,6 +67,10 @@ const openEditorApp = ({ useSsl = true } = {}) => ({
                 getMessage("OpenEditor_Client_Console_Fatal_Error")
               )
             );
+          },
+          [clientMessages.DECODE_ERROR_MESSAGE]: path => {
```
Comment (dor-itzhaki) [inline packages/corvid-cli/src/apps/open-editor.js]: there is no test for this scenario

File: packages/corvid-local-server/src/socket-api/index.js [hunk 12]
```diff
@@ -48,9 +48,13 @@ const initServerApi = (
   const isEditorConnected = () =>
     Object.keys(editorServer.io.sockets.connected).length > 0;
 
-  const getSiteDocument = () => {
-    return localSite.getSiteDocument();
-  };
+  const getSiteDocument = () =>
+    localSite.getSiteDocument().catch(e => {
+      if (e.name === "ParseError") {
```
Comment (dor-itzhaki) [inline packages/corvid-local-server/src/socket-api/index.js]: why not "e typeof ParseError" ?

File: packages/corvid-local-site/src/siteConverter.js [hunk 13]
```diff
@@ -199,13 +201,19 @@ const editorDocumentToLocalDocumentFiles = siteDocument => {
 };
 
 const localDocumentFilesToEditorDocument = localDocumentFiles => {
-  const parsedLocalDocumentFiles = localDocumentFiles.map(file =>
-    isPathOfWixFile(file.path)
-      ? Object.assign({}, file, {
+  const parsedLocalDocumentFiles = localDocumentFiles.map(file => {
+    if (isPathOfWixFile(file.path)) {
+      try {
```
Comment (dor-itzhaki) [inline packages/corvid-local-site/src/siteConverter.js]: extract to a function ?

File: packages/corvid-cli/test/it/edit.spec.js (hunk 9 above)
Comment (kostagersh) [reply packages/corvid-cli/test/it/edit.spec.js]: done

File: packages/corvid-local-site/src/siteConverter.js (hunk 13 above)
Comment (kostagersh) [reply packages/corvid-local-site/src/siteConverter.js]: done

File: packages/corvid-local-server/src/socket-api/index.js (hunk 12 above)
Comment (kostagersh) [reply packages/corvid-local-server/src/socket-api/index.js]: done

File: packages/corvid-cli/test/it/edit.spec.js (hunk 10 above)
Comment (kostagersh) [reply packages/corvid-cli/test/it/edit.spec.js]: done

Comment (kostagersh) [reply packages/corvid-cli/test/it/edit.spec.js]: reverted

Comment (kostagersh) [reply packages/corvid-cli/test/it/edit.spec.js]: there is no such option now, tried to create, but it was clumsy

File: packages/corvid-cli/src/apps/open-editor.js (hunk 11 above)
Comment (kostagersh) [reply packages/corvid-cli/src/apps/open-editor.js]: done
